src/App.tsx

```diff
@@ -1,204 +1,238 @@
+        setIsLoading(currentLoads =>
+          currentLoads.filter(loadId => loadId !== todoId),
+        );
+      });
+  };
+
+  const handleDeleteCompleted = () => {
+    Promise.allSettled(completedTodosId.map(todoId => handleDelete(todoId)));
+  };
+
+  const handleStatusChange = (updatedTodo: Todo) => {
+    setIsLoading(currentLoads => [...currentLoads, updatedTodo.id]);
+
+    return updateTodo({ ...updatedTodo, completed: !updatedTodo.completed })
+      .then(() => {
+        setTodos(currentTodos => {
+          return currentTodos.map(todo =>
+            todo.id === updatedTodo.id
+              ? { ...todo, completed: !updatedTodo.completed }
+              : todo,
+          );
+        });
+      })
+      .catch(() => {
+        setErrorMessage('Unable to update a todo');
+      })
+      .finally(() => {
+        setIsLoading(currentLoads =>
+          currentLoads.filter(loadId => loadId !== updatedTodo.id),
+        );
+      });
+  };
+
+  const handleToggleAll = () => {
+    const todosToToggle =
+      completedTodosId.length === todos.length
+        ? todos
+        : todos.filter(todo => !todo.completed);
+
+    Promise.allSettled(todosToToggle.map(todo => handleStatusChange(todo)));
+  };
+
+  const handleEditTodo = (updatedTodo: Todo) => {
+    setIsLoading(currentLoads => [...currentLoads, updatedTodo.id]);
+
+    return updateTodo(updatedTodo)
+      .then(() => {
+        setTodos(currentTodos => {
+          return currentTodos.map(todo =>
+            todo.id === updatedTodo.id ? updatedTodo : todo,
+          );
+        });
+        setIsUpdateError(false);
+      })
+      .catch(() => {
+        setIsUpdateError(true);
+        setErrorMessage('Unable to update a todo');
+      })
+      .finally(() => {
+        setIsLoading(currentLoads =>
+          currentLoads.filter(loadId => loadId !== updatedTodo.id),
+        );
+      });
+  };
+
   return (
     <div className="todoapp">
       <h1 className="todoapp__title">todos</h1>
 
       <div className="todoapp__content">
-        <header className="todoapp__header">
-          {/* this button should have `active` class only if all todos are completed */}
-          <button
-            type="button"
-            className="todoapp__toggle-all active"
-            data-cy="ToggleAllButton"
+        <Header
+          textField={textField}
+          onTextField={handleTextField}
+          onSubmit={handleSubmit}
+          isSubmiting={isSubmiting}
+          field={field}
+          onToggleAll={handleToggleAll}
+          isToggleActive={completedTodosId.length !== 0}
+          isToggleVisible={!!todosLength}
+        />
+
+        <TodoList
+          visibleTodos={visibleTodos}
+          onDelete={handleDelete}
+          isLoading={isLoading}
+          tempTodo={tempTodo}
+          textField={textField}
+          onStatusChange={handleStatusChange}
+          onEdit={handleEditTodo}
+          isUpdateError={isUpdateError}
+        />
```

Comment:
You can also render the list of todos this way. Currently, TodoList is passing a lot of props directly to the TodoItem component. You could simplify it by flattening the props as suggested.
```suggestion
        <TodoList tempTodo={tempTodo}>
          {visibleTodos.length !== 0 && (
            <>
              {visibleTodos.map(todo => (
                <TodoItem
                  key={todo.id}
                  todo={todo}
                  onDelete={handleDelete}
                  isLoading={isLoading}
                  onStatusChange={handleStatusChange}
                  onEdit={handleEditTodo}
                  isUpdateError={isUpdateError}
                />
              ))}
            </>
          )}
        </TodoList>
```